### DISCUSSION TITLE
How can I calculate "probability" for a result segment?

### DISCUSSION THREAD

**Discussion by toanhn** [SYSTRAN/faster-whisper#1043] [Q&A]
I see in result segments, each word has the value "probability", but in each segment, I do not see the value "probability".
How can I calculate "probability" for a result segment?
Thanks

**Comment by phineas-pta** [comment]
it's complicated

**Comment by MahmoudAshraf97** [comment]
you need to define what does the probability for each segment means or represents before you ask how to calculate it

**Reply by dgoryeo** [reply]
I guess there is no one-size-fits-all solution and each language or context may have its own characteristics. I found the question useful/relevant, however I have no subject matter experise so I asked GPT-o1-preview for possible approaches. FWIW, here is a summary of the output: 

`
Here are some scientifically sound methods to compute the segment probability:

1. Product of Probabilities (Joint Probability)
Method: Calculate the joint probability of the segment by multiplying the probabilities of all words in the segment.

2. Sum of Log Probabilities
Method: Convert word probabilities to log probabilities, sum them up, and use the total log probability as the segment's confidence score.

3. Average Log Probability
Method: Calculate the average of the log probabilities of the words in the segment.

4. Geometric Mean of Probabilities
Method: Calculate the geometric mean of the word probabilities.

`

**Comment by gongouveia** [comment]
@toanhn @phineas-pta It is not complicated, in the output of the inference, you have access to the logits of each word, with this logit you can convert it to a probability of each word (type of confidence). Mind that  the logit of the probability is the logarithm of the odds.
Now you can do the average of the probabilities :)

Let me know if this message was helpful.

**Reply by toanhuynhnguyen** [reply]
Thanks bro

**Comment by arcticloud** [comment]
Each `Segment` already carries `avg_logprob` — the average token log-probability across that segment. Convert it to a 0–1 confidence with `exp`:

```python
import math
for seg in segments:
    confidence = math.exp(seg.avg_logprob)   # ~0..1, higher = more confident
```

Two related fields worth using:
- `seg.no_speech_prob` — probability the segment is silence/non-speech. A high value here is the best signal for a hallucinated/garbage segment.
- If you run with `word_timestamps=True`, each word has `.probability`, so you can compute a finer segment score yourself: `sum(w.probability for w in seg.words) / len(seg.words)`.

One caveat: neither `exp(avg_logprob)` nor the word probabilities are *calibrated* probabilities — they're confidence proxies. They're great for relative ranking and for thresholding out bad segments (e.g. drop segments where `exp(avg_logprob) < 0.4` or `no_speech_prob > 0.6`), but don't read them as a literal "X% correct".
